src_zlib/service/zlib_service.cpp

```diff
@@ -214,6 +234,163 @@ class probe
 
 //------------------------------------------------
 
+
+namespace {
+void* zalloc_impl(
+    void* /* opaque */,
+    unsigned items,
+    unsigned size)
+{
+    try {
+        return ::operator new(items * size);
+    } catch(std::bad_alloc const&) {
+        return Z_NULL;
+    }
+}
+
+void zfree_impl(void* /* opaque */, void* addr)
+{
+    ::operator delete(addr);
+}
+} // namespace
+
+class BOOST_HTTP_PROTO_ZLIB_DECL
+    zlib_filter final : public filter
+{
+private:
+    z_stream stream_;
+
+    void init(bool use_gzip);
+
+public:
+    zlib_filter(bool use_gzip = false);
+    ~zlib_filter();
+
+    zlib_filter(zlib_filter const&) = delete;
+    zlib_filter& operator=(zlib_filter const&) = delete;
+
+    filter::results
+    on_process(
+        buffers::mutable_buffer out,
+        buffers::const_buffer in,
+        bool more) override;
+};
+
+zlib_filter::
+zlib_filter(bool use_gzip)
+{
+    stream_.zalloc = &zalloc_impl;
+    stream_.zfree = &zfree_impl;
+    stream_.opaque = nullptr;
+    init(use_gzip);
+}
+
+zlib_filter::
+~zlib_filter()
+{
+    deflateEnd(&stream_);
+}
+
+void
+zlib_filter::
+init(bool use_gzip)
+{
+    int ret = -1;
+
+    int window_bits = 15;
+    if( use_gzip )
+        window_bits += 16;
+
+    int mem_level = 8;
+
+    ret = deflateInit2(
+        &stream_, Z_DEFAULT_COMPRESSION, Z_DEFLATED,
+        window_bits, mem_level, Z_DEFAULT_STRATEGY);
+
+    if( ret != Z_OK )
+        throw ret;
+
+    stream_.next_out = nullptr;
+    stream_.avail_out = 0;
+
+    stream_.next_in = nullptr;
+    stream_.avail_in = 0;
+}
+
+filter::results
+zlib_filter::
+on_process(
+    buffers::mutable_buffer out,
+    buffers::const_buffer in,
+    bool more)
```

Comment:
You need to remember that `more` because `true`, because `on_process` may need to be called later for the case where there is no more input, but the amount of buffered output is larger than the provided mutable buffer. In this case we have to assert that when `more` is first passed as `true`, subsequent calls do not have `more==false`. This needs to be documented in the `filter::on_process` member javadoc.

Reply:
I'm not sure I understand.
The flow should be `true -> false` for more. The code that works now does:
```cpp
auto flush = more ? Z_NO_FLUSH : Z_FINISH;
```
i.e. when `more` is true, we don't invoke deflate() with any flushing. But if more is false, we know there's no more pending input so we tell zlib to go ahead and start flushing output.

I'm not sure about the `filter::on_process` thing either. Examining it more closely, this seems like it's meant to be a public type but I'm not sure any of this should be public. If you notice, I've had to make the serializer a friend of it.

Reply:
Yes of course you are right, I got true and false reversed, but my point still stands.